cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternalsCommon.qll

```diff
@@ -103,10 +103,18 @@ private class PointerOrArrayOrReferenceType extends Cpp::DerivedType {
  * (i.e., `countIndirections(e.getUnspecifiedType())`).
  */
 private int countIndirections(Type t) {
-  result = any(Indirection ind | ind.getType() = t).getNumberOfIndirections()
-  or
-  not exists(Indirection ind | ind.getType() = t) and
-  result = 0
+  // We special case void pointers because we don't know how many indirections
+  // they really have. In a Glorious Future we could do a pre-analysis to figure out
+  // which kinds of values flows into the type and use the maximum number of
+  // indirections flowinginto the type.
+  if t instanceof Cpp::VoidPointerType
+  then result = 2
```

Comment:
Why 2?

Reply:
Given a variable `void* p` we currently, we have a dataflow node for `p` and for `*p` with types `void*` and `void` respectively (we should probably have a better type for `*p` eventually, but that's left as future work). This means that we'll have flow in cases such as:
```cpp
int x = source(); // x is tainted
void* p = (void*)&x; // *p is tainted
sink(*(int*)p);
```
since there'll be flow from `source() -> x -> indirection of p -> *(int*)p`.

That's all well and good. Now, consider this slightly different program (that is equivalent to the situation added in https://github.com/github/codeql/pull/12316/commits/fddc54f4d26372212dc4cc147cb9ca6ef4cf0bff):
```cpp
char* p = indirect_source(); // *p is tainted
void* q = (void*)&p; // **q is tainted
sink(**(char**)q);
```
prior to this PR we wouldn't get flow in this case because we didn't have a dataflow node for `**q` (since `q` has type `void*` and thus we only created dataflow nodes for `q` and `*q`). This negatively impacted Samate quite a bit.

In the future I would we can do some pre analysis that decides how many dataflow nodes we need to allocate for a given void pointer (since, unlike with other pointer types, it's not clear how many possible dereferences people intend to do on it). But for now, we regain all the Samate results (without impacting performance as far as I can see) by just saying "we create a dataflow node for `q`, `*q` and `**q` for each void pointer `q`).

Does that make sense?

Reply:
I think I understand - we already had indirections 0 and 1, it's 2 because we want to support `**q` as that sort of thing is quite common.  And it could be `[2, 3]` instead to support `***q` as well but we have to stop somewhere.

Reply:
Yep, exactly.